### DISCUSSION TITLE
Fix https://github.com/vercel/turbo/pull/8137

### DISCUSSION THREAD

**Discussion by mctrafik** [vercel/turborepo#8524] [Ideas]
### Goals

To let users avoid writing full package scope.

### Non-goals

Break existing workflows

### Background

This PR is pretty unpopular: https://github.com/vercel/turbo/pull/8137

### Proposal

Bring back the scope detection functionality. Make it better though, so that conflicts have a deterministic behavior.

**Comment by echocrow** [comment]
+1 to bring back namespace inference. while not critical, it was convenient to not have to spell out the same namespace/scope every time. this also aligned with [pnpm's `--filter`](https://pnpm.io/filtering). for example:
- `pnpm dev --filter=website`: runs `@my-app/website#dev` via turborepo (v1), following the usual task dependencies, cache, etc.
- `pnpm --filter=website dev`: directly runs `@my-app/website#dev` via pnpm (bypasses turbo)

That said, I do see how the original implementation—particularly with a zero exit code—is potentially problematic. Off the top of my head, when multiple packages match a non-wildcard'd name, I'd assume turbo to either invoke all matching packages, or error out with a non-zero exit code.

Eating my own words: I checked to see how `pnpm` handles this scenario; it too actually aborts with a zero code, and prints the message _"No projects matched the filters in /path/to/repo"_. 🥲

**Comment by anthonyshew** [comment]
Hey, thanks for the idea! 8137 is the fix for many ambiguous behaviors through `turbo`.

I can understand how, at a surface level, this change feels inconvenient, but it delivers a more consistent experience.
- Package managers have ambiguous behavior around their filtering commands (as @echocrow just found out). 😄 So, we chose to index on explicit correctness.
- Package managers also have ambiguous behavior around how they handle the scope inference in how they build their package graphs. Using the namespace removes an entire class of error from `turbo`.
- It matches what you'd write in `turbo.json` and other places around `turbo` where you'd have to use the namespace. A common failure mode we were seeing was users trying to use un-namespaced package names in their `turbo.json` and other places and not understanding why it wouldn't work, since they were using the un-namespaced version to write`--filter`s.

Overall, it's a tradeoff between correctness and convenience, and this is a place where correctness is the right call. You also may be interested in [Automatic Package Scoping](https://turbo.build/repo/docs/crafting-your-repository/running-tasks#automatic-package-scoping) if you're not leveraging that feature yet, so you don't have to type any filter at all in many cases.

**Reply by echocrow** [reply]
Cheers for indulging and sharing all the extra context! It sounds like you've seen more issues arise from the inference, so this feature was more trouble than it's worth.

Speaking from experience, I can see more correctness also help out newcomers. When we first adopted turbo a while ago, I remember being confused why `pnpm build --filter=my-pkg` would build `@scope/my-pkg`, but not e.g. `@scope/my-pkg-extras`. However, `--filter=my-pkg*` somehow matched nothing, while `--filter=@scope/my-pkg*` matched both. Took some time to realize it was not a partial match, but either an exact match or "single-auto-scope" match, and wildcards (I believe) would not work with inferred scopes. Go figure.

Bit unfortunate to see it go, but we'll get used to it in no time (especially once command history catches up).

**Comment by mctrafik** [comment]
I feel like your reply misses the point, @anthonyshew. If you're going to take away a convenience that already exists, you should replace it with something.

I.e.:

1. add a flag to infer on turbo command
2. allow specifying a project name alias/es in turbo.json
3. Allow setting a config to turn on inference by default

I understand that taking away functinoality reduces bugs.. that's how deleting code works in general and I it's not something that should be celebrated. I feel like this discussion isn't talking about what you've `taken away` from users.

**Reply by anthonyshew** [reply]
I understand that the functionality was removed. That was the goal and why it was a part of a major upgrade. 😄

The more important thing is that now, despite the inconvenience, the behavior is more correct, according to the user feedback we'd received. We knew that there would be a group that didn't appreciate the change, as well, and I understand why. I wish we could make everyone happy at all times, but this is the sort of thing that we have to asses the tradeoffs and make a decision for the betterment of the project.

We tested this on ourselves and users who gave us feedback on pre-release canaries. We all used it for long enough before we realized it was the right choice. We did go through a period of being mildly annoyed but, after adjusting our habits, figured out that it's a good change for the long-term.

As far as the suggestions you've made go, I'm not sure that they're viable given that they don't answer to the original reasons for the change. Additionally, there'd be added overhead on developers to know if you _could_ use the namespace inference depending on which Turborepo you're in.
